File: 2018/1/day1.pl
#!/usr/bin/env perl

use strict;
use warnings;

sub change_freq {
    my ($freq, $shift) = @_;
    $shift = int($shift);
    return $freq + $shift;
}

sub part_one {
    my @input = @_;
    my $freq = 0;
    foreach my $shift (@input) {
        $freq = change_freq($freq, $shift);
    }
    return $freq;
}

sub part_two {
    my @input = @_;
    my $freq = 0;
    my %seen_freqs;
    for (;;) {
        foreach my $shift (@input) {
            $freq = change_freq($freq, $shift);
            $seen_freqs{$freq}++;
            return $freq if ($seen_freqs{$freq} == 2);
        }
    }
}

my @input = <STDIN>;
print part_one(@input), "\n";
print part_two(@input), "\n";

File: 2018/1/day1.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

func changeFrequency(curFreq int, shift string) int {
	// Convert the string to an int
	// Atoi can deal with turning +1 into 1
	shiftInt, err := strconv.Atoi(shift)
	if err != nil {
		return 0
	}
	return curFreq + shiftInt

}

func partOne(input []string) int {
	freq := 0
	for _, shift := range input {
		freq = changeFrequency(freq, shift)
	}
	return freq
}

func partTwo(input []string) int {
	seenFreqs := make(map[int]int)
	freq := 0
	// Fixes the bug for the +1 -1 test - we were told it starts at 0 therefore it's already been "seen"
	seenFreqs[0] = 1
	for {
		for _, shift := range input {
			freq = changeFrequency(freq, shift)
			seenFreqs[freq]++
			if seenFreqs[freq] == 2 {
				return freq
			}
		}
	}
}

func main() {
	var input []string
	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		input = append(input, scanner.Text())
	}
	fmt.Println("Part one:", partOne(input))
	fmt.Println("Part two:", partTwo(input))
}
